lib/chef/resource/user/mac_user.rb

```diff
@@ -0,0 +1,105 @@
+#
+# Author:: Ryan Cragun (<ryan@chef.io>)
+# Copyright:: Copyright 2019, Chef Software Inc.
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require_relative "../user"
+
+class Chef
+  class Resource
+    class User
+      # Provide a user resource that is compatible with default TCC restrictions
+      # that were introduced in macOS 10.14.
+      #
+      # Changes:
+      #
+      # * This resource and the corresponding provider have been modified to
+      #   work with default macOS TCC policies. Direct access to user binary
+      #   plists are no longer permitted by default, thus we've chosen to use
+      #   use newer methods of creating, modifying and removing users.
+      #
+      # * Due to the tooling required by the provider this provider is only
+      #   suitable for use on macOS >= 10.14. Support for older platforms has
+      #   been removed.
+      #
+      # New Features:
+      #
+      # * Primary group management is now included.
+      #
+      # * 'admin' is now a boolean property that configures a user to an admin.
+      #
+      # * 'admin_username' and 'admin_password' are new properties that define the
+      #   admin user credentials required for toggling SecureToken for an
+      #   exiting user.
+      #
+      #   The 'admin_username' must correspond to a system admin with SecureToken
+      #   enabled in order to toggle SecureToken.
+      #
+      # * 'secure_token' is a boolean property that sets the desired state
+      #   for SecureToken. SecureToken token is required for FileVault full
+      #   disk encryption.
+      class MacUser < Chef::Resource::User
+        resource_name :mac_user
+
+        provides :mac_user
+        provides :user, os: "darwin", platform_version: ">= 10.14"
+
+        property :iterations, Integer,
+          description: "macOS platform only. The number of iterations for a password with a SALTED-SHA512-PBKDF2 shadow hash.",
+          default: 57803, desired_state: false
+
+        # Overload gid so we can set our default. NilClass is for backwards compat
+        # and 20 is the macOS "staff" group.
+        property :gid, [String, Integer, NilClass], description: "The numeric group identifier.", default: 20, coerce: ->(gid) do
```

Comment:
It looks like this is `property :gid, Integer` since you can't assign a nil and strings get coerced, and the validation should apply to after all coercion.

Reply:
I allowed `String` because if group lookup failed during compile time then we'd allow try resolving the group again in the provider in case a `group` resource had converged and created the group we want before our `user` resource.

Seems to me the issue is that previously we didn't do group management, so if GID resolution happened once and failed we could raise an error because a group _should_ have been created before the user which you're trying to associate it with. But now we do group management so we have to assume a user is giving us the GID or group name they want us to handle, regardless of whether or not it exists at compile time or converge time.

I've convinced myself that regardless or whether or not we want to keep group management, the property probably need to be both an `Integer` or `String` and that we should only resolve it a single time in the provider, not with coercion. The reason we'll want to do it later is to handle the case where a prior `group` resource creates a group that doesn't exist at compile time. If we used coercion to resolve it at compile time that would surely fail.

We have two options after that:

* get rid of group management. Try and resolve the group name in the provider and raise an error if it doesn't exist, associate the primary ID.
* keep group management, create the group if it doesn't resolve, append the user to the group, associate the primary ID.

I went with the second option.

Reply:
cool that makes sense.